src/main/java/org/opensearch/neuralsearch/query/HybridBulkScorer.java

```diff
@@ -0,0 +1,204 @@
+        this.maxDoc = maxDoc;
+        this.hybridQueryDocIdStream = new HybridQueryDocIdStream(this);
+    }
+
+    @Override
+    public int score(LeafCollector collector, Bits acceptDocs, int min, int max) throws IOException {
+        collector.setScorer(hybridSubQueryScorer);
+        // making sure we are not going over the global limit defined by maxDoc
+        max = Math.min(max, maxDoc);
+        // advance all scorers to the segment's minimum doc id
+        int[] docsIds = advance(min, scorers);
+        while (allDocIdsUsed(docsIds, max) == false) {
+            scoreWindow(collector, acceptDocs, min, max, docsIds);
+        }
+        return getNextDocIdCandidate(docsIds);
+    }
+
+    private void scoreWindow(LeafCollector collector, Bits acceptDocs, int min, int max, int[] docIds) throws IOException {
+        // pick the lowest out of all not yet used doc ids
+        int topDoc = -1;
+        for (int docId : docIds) {
+            if (docId < max) {
+                topDoc = docId;
+                break;
+            }
+        }
+
+        final int windowBase = topDoc & ~MASK; // take the next match (at random) and find the window where it belongs
+        final int windowMin = Math.max(min, windowBase);
+        final int windowMax = Math.min(max, windowBase + WINDOW_SIZE);
+        // collect doc ids and scores for this window using leaf collector
+        scoreWindowIntoBitSetWithSubqueryScorers(collector, acceptDocs, max, docIds, windowMin, windowMax, windowBase);
+    }
+
+    /**
+     * Collect scores for the window using segment level leaf collector
+     * @param collector leaf collector for the segment
+     * @param acceptDocs bitset with live docs
+     * @param max max doc id
+     * @param docIds last used doc ids per scorer
+     * @param windowMin min doc id of this collector window
+     * @param windowMax max doc id of this collector window
+     * @param windowBase offset for this collector window
+     * @throws IOException
+     */
+    private void scoreWindowIntoBitSetWithSubqueryScorers(
+        LeafCollector collector,
+        Bits acceptDocs,
+        int max,
+        int[] docIds,
+        int windowMin,
+        int windowMax,
+        int windowBase
+    ) throws IOException {
+        for (int subQueryIndex = 0; subQueryIndex < scorers.length; subQueryIndex++) {
+            if (Objects.isNull(scorers[subQueryIndex]) || docIds[subQueryIndex] >= max) {
+                continue;
+            }
+            DocIdSetIterator it = scorers[subQueryIndex].iterator();
+            int doc = docIds[subQueryIndex];
+            if (doc < windowMin) {
+                doc = it.advance(windowMin);
+            }
+            while (doc < windowMax) {
+                if (Objects.isNull(acceptDocs) || acceptDocs.get(doc)) {
+                    int d = doc & MASK;
+                    if (needsScores) {
+                        float score = scorers[subQueryIndex].score();
```

Comment:
Question for learn: does the cost calculation happens only once during initialization?

Reply:
yes, the contract for cost() method is quite relaxed, so it most scorers it's calculated once to save resources. Here is from [Lucene doc](https://github.com/apache/lucene/blob/branch_10_0/lucene/core/src/java/org/apache/lucene/search/DocIdSetIterator.java):
```
Returns the estimated cost of this DocIdSetIterator.
This is generally an upper bound of the number of documents this iterator might match, but may be a rough heuristic, hardcoded value, or otherwise completely inaccurate.
```